lib/matplotlib/backend_tools.py

```diff
@@ -0,0 +1,928 @@
+        if self.navigation.messagelock.locked():
+            return
+
+        message = ' '
+
+        if event.inaxes and event.inaxes.get_navigate():
+            try:
+                s = event.inaxes.format_coord(event.xdata, event.ydata)
+            except (ValueError, OverflowError):
+                pass
+            else:
+                message = s
+        self.navigation.message_event(message, self)
+
+
+class RubberbandBase(ToolBase):
+    """Draw and remove rubberband"""
+    def trigger(self, sender, event, data):
+        """Call `draw_rubberband` or `remove_rubberband` based on data"""
+        if not self.figure.canvas.widgetlock.available(sender):
+            return
+        if data is not None:
+            self.draw_rubberband(*data)
+        else:
+            self.remove_rubberband()
+
+    def draw_rubberband(self, *data):
+        """Draw rubberband
+
+        This method must get implemented per backend
+        """
+        pass
+
+    def remove_rubberband(self):
+        """Remove rubberband
+
+        This method must get implemented per backend
+        """
+        pass
+
+
+class ToolQuit(ToolBase):
+    """Tool to call the figure manager destroy method"""
+
+    description = 'Quit the figure'
+    keymap = rcParams['keymap.quit']
+
+    def trigger(self, sender, event, data=None):
+        Gcf.destroy_fig(self.figure)
+
+
+class ToolEnableAllNavigation(ToolBase):
+    """Tool to enable all axes for navigation interaction"""
+
+    description = 'Enables all axes navigation'
+    keymap = rcParams['keymap.all_axes']
+
+    def trigger(self, sender, event, data=None):
+        if event.inaxes is None:
+            return
+
+        for a in self.figure.get_axes():
+            if event.x is not None and event.y is not None \
+                    and a.in_axes(event):
+                a.set_navigate(True)
+
+
+class ToolEnableNavigation(ToolBase):
+    """Tool to enable a specific axes for navigation interaction"""
+
+    description = 'Enables one axes navigation'
+    keymap = (1, 2, 3, 4, 5, 6, 7, 8, 9)
+
+    def trigger(self, sender, event, data=None):
+        if event.inaxes is None:
+            return
+
+        n = int(event.key) - 1
+        for i, a in enumerate(self.figure.get_axes()):
+            # consider axes, in which the event was raised
+            # FIXME: Why only this axes?
+            if event.x is not None and event.y is not None \
+                    and a.in_axes(event):
+                    a.set_navigate(i == n)
+
+
+class ToolGrid(ToolToggleBase):
+    """Tool to toggle the grid of the figure"""
+
+    description = 'Toogle Grid'
+    keymap = rcParams['keymap.grid']
+
+    def trigger(self, sender, event, data=None):
+        if event.inaxes is None:
+            return
+        ToolToggleBase.trigger(self, sender, event, data)
+
+    def enable(self, event):
+        event.inaxes.grid(True)
+        self.figure.canvas.draw()
```

Comment:
with respect to key_press_handler, it isn't implementing a method called
"mpl_connect". It is using the mpl_connect because the default keypress
handler is getting connected to the canvas's callback registry. This is a
different callback registry. Elsewhere in mpl, we have widgets and other
things that instantiate their own callback registry, and they do not create
a method called mpl_connect() to add to it. I would just go with
"connect()", but "nav_connect()" is fine, too.

On Wed, Jan 21, 2015 at 3:06 PM, Federico Ariza <notifications@github.com>
wrote:

> In lib/matplotlib/backend_tools.py
> <https://github.com/matplotlib/matplotlib/pull/3652#discussion_r23329795>:
>
> > +
> > +
> > +class ToolGrid(ToolToggleBase):
> > +    """Tool to toggle the grid of the figure"""
> > +
> > +    description = 'Toogle Grid'
> > +    keymap = rcParams['keymap.grid']
> > +
> > +    def trigger(self, sender, event, data=None):
> > +        if event.inaxes is None:
> > +            return
> > +        ToolToggleBase.trigger(self, sender, event, data)
> > +
> > +    def enable(self, event):
> > +        event.inaxes.grid(True)
> > +        self.figure.canvas.draw()
>
> No problem, I'll change that, it is this way in key_press_handler
> implementation, all I did was a (copy/paste)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/3652/files#r23329795>.
>